experimental/oauther/oauth2.go

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) 2019-present Mattermost, Inc. All Rights Reserved.
+// See License for license information.
+
+package oauther
+
+import (
+	"encoding/json"
+
+	"github.com/gorilla/mux"
+	"github.com/mattermost/mattermost-plugin-api/experimental/bot/logger"
+	"github.com/mattermost/mattermost-server/v5/plugin"
+	"golang.org/x/oauth2"
+)
+
+const (
+	oAuth2StateTimeToLive = 300 // seconds
+)
+
+type OAuther interface {
+	GetToken(userID string) (*oauth2.Token, error)
+	GetURL() string
+	Deauth(userID string) error
+}
+
+type oAuther struct {
+	api             plugin.API
+	onConnect       func(userID string, token *oauth2.Token)
+	storePrefix     string
+	pluginURL       string
+	oAuthURL        string
+	connectedString string
+	config          *oauth2.Config
+	logger          logger.Logger
+}
+
+func NewOAuther(
+	r *mux.Router,
+	api plugin.API,
+	pluginURL,
+	oAuthURL,
```

Comment:
Would it make sense to have a sane default for this instead of requiring it to be configurable?

Reply:
We could use a function like:
`NewOAuther(..., options [string]interface{})`
If the option is not there, just add the default. Nevertheless, I am not specially a fan of the options maps (hard to guess what options you can pass).
Also, I prefer to require this configuration, to avoid clashes. If I set the oAuthURL, I know I cannot do anything with those URLs. The same for KVStore prefix. If some errors start to pop up because of the default values, it is harder to track the problem. Better safe than sorry, I think.

Reply:
@lieut-data functional options to the rescue? :)

Reply:
I have made a take on functional options. If they look nice, I may start using them on other libraries.